### ISSUE TITLE
futures should have an `always` or something

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#665]
often you need to do some cleanup that needs to always happen, why not

```swift
extension EventLoopFuture {
    func always(_ body: () -> Void) -> EventLoopFuture<T>
}
```

that always runs?

**Comment by Lukasa** [comment]
Why does `always` return a `T`?

**Comment by weissi** [comment]
because it's too early in the morning 😂. fixing it :P

**Comment by normanmaurer** [comment]
@weissi so this is something that will be executed no matter if the future is completed because of a failure or success ?

**Comment by Lukasa** [comment]
@weissi Is this any different to:

```swift
public func whenComplete(_ callback: @escaping () -> Void)
```

Because we already have that.

**Comment by Lukasa** [comment]
Oh, I see, it chains. Sure, why not.

**Comment by normanmaurer** [comment]
If we do this then we should most likely just mark `whenComplete` as deprecated and introduce `always` but mark its return value as `@discardableResult`.

**Comment by weissi** [comment]
@normanmaurer need both chaining and non-chaining variants

**Comment by normanmaurer** [comment]
@weissi why ?

**Comment by weissi** [comment]
@normanmaurer because otherwise you can't write this:

```swift
func runDatabaseQuery(_ query: String, database: DB) -> EventLoopFuture<[Row]> {
    return database.openConnetion { conn in
        conn.runQuery(query).always {
            conn.close()
        }
    }
}
```

You want to always close the DB connection. Currently you need to spell this

```swift
func runDatabaseQuery(_ query: String, database: DB) -> EventLoopFuture<[Row]> {
    return database.openConnetion { conn in
        let f = conn.runQuery(query)
        f.whenComplete {
            conn.close()
        }
        return f
    }
}
```

which is super ugly

**Comment by normanmaurer** [comment]
@weissi I may need more coffee but I still dont get why we could not only have `always` at the end ?

**Comment by weissi** [comment]
@Lukasa we'll still create an extra, unnecessary future. We also have `then` & `whenSuccess`, `thenIfError` & `whenFailure`, but _nothing_ & `whenSuccess`. I just want to fill the _nothing_-hole :)

**Comment by normanmaurer** [comment]
@weissi why ? Couldn't it just return `self` in this case ?

**Comment by weissi** [comment]
@normanmaurer true

**Comment by Lukasa** [comment]
Bear in mind that the semantics of that are *slightly* different than the semantics of returning a new future, though only slightly.

**Comment by normanmaurer** [comment]
@Lukasa but does it matter at all ?

**Comment by Lukasa** [comment]
Depends if you're trying to be really certain about future execution. *Probably* not though.

**Comment by Palleas** [comment]
@normanmaurer @Lukasa @weissi Is that something we'd still want to add? 🤓

**Comment by weissi** [comment]
I'm still very much in favour of having this, every time I use NIO as a user I want to have it :). Sure, it's pretty straightforward to work around but I often have something I need to close, a connection, a file, whatever and of course I want to do it _always_ if there's an error or not.

**Comment by weissi** [comment]
Fixed by #981
